internal/tasks/tasks.go

```diff
@@ -98,25 +102,106 @@ func (j *biosResetTask) Asset() *model.Asset {
+
+func NewTask(task *rctypes.Task[any, any]) (*RCTask, error) {
+	paramsJSON, ok := task.Parameters.(json.RawMessage)
+	if !ok {
+		return nil, errInvalidConditionParams
+	}
+
+	params := rctypes.BiosControlTaskParameters{}
+	if err := json.Unmarshal(paramsJSON, &params); err != nil {
+		return nil, err
+	}
+
+	// deep copy fields referenced by pointer
+	asset, err := copystructure.Copy(task.Server)
+	if err != nil {
+		return nil, errors.Wrap(errTaskConv, err.Error()+": Task.Server")
+	}
+
+	fault, err := copystructure.Copy(task.Fault)
+	if err != nil {
+		return nil, errors.Wrap(errTaskConv, err.Error()+": Task.Fault")
+	}
+
+	return &RCTask{
+		StructVersion: task.StructVersion,
+		ID:            task.ID,
+		Kind:          task.Kind,
+		State:         task.State,
+		Status:        task.Status,
+		Parameters:    &params,
+		Fault:         fault.(*rctypes.Fault),
+		FacilityCode:  task.FacilityCode,
+		Server:        asset.(*rtypes.Server),
+		WorkerID:      task.WorkerID,
+		TraceID:       task.TraceID,
+		SpanID:        task.SpanID,
+		CreatedAt:     task.CreatedAt,
+		UpdatedAt:     task.UpdatedAt,
+		CompletedAt:   task.CompletedAt,
+	}, nil
+}
+
+func (task *RCTask) ToGeneric() (*rctypes.Task[any, any], error) {
+	paramsJSON, err := task.Parameters.Marshal()
+	if err != nil {
+		return nil, errors.Wrap(errTaskConv, err.Error()+": Task.Parameters")
+	}
+
+	// deep copy fields referenced by pointer
+	asset, err := copystructure.Copy(task.Server)
+	if err != nil {
+		return nil, errors.Wrap(errTaskConv, err.Error()+": Task.Server")
+	}
+
+	fault, err := copystructure.Copy(task.Fault)
+	if err != nil {
+		return nil, errors.Wrap(errTaskConv, err.Error()+": Task.Fault")
+	}
+
+	return &rctypes.Task[any, any]{
+		StructVersion: task.StructVersion,
+		ID:            task.ID,
+		Kind:          task.Kind,
+		State:         task.State,
+		Status:        task.Status,
+		Parameters:    paramsJSON,
+		Fault:         fault.(*rctypes.Fault),
+		FacilityCode:  task.FacilityCode,
+		Server:        asset.(*rtypes.Server),
+		WorkerID:      task.WorkerID,
+		TraceID:       task.TraceID,
+		SpanID:        task.SpanID,
+		CreatedAt:     task.CreatedAt,
+		UpdatedAt:     task.UpdatedAt,
+		CompletedAt:   task.CompletedAt,
+	}, nil
+}
+
 // TaskRunner Will run the task by executing the individual steps in the task,
 // and reports task status using the publisher.
 type TaskRunner struct {
-	publisher  controller.ConditionStatusPublisher
-	task       Task
+	publisher  ctrl.Publisher
+	task       *RCTask
+	oldTask    Task
 	taskStatus *TaskStatus
 }
 
 // NewTaskRunner creates a TaskRunner to run a specific Task
-func NewTaskRunner(publisher controller.ConditionStatusPublisher, task Task) *TaskRunner {
+func NewTaskRunner(publisher ctrl.Publisher, oldTask Task, task *RCTask) *TaskRunner {
 	return &TaskRunner{
 		publisher:  publisher,
 		task:       task,
-		taskStatus: NewTaskStatus(task.Name(), rctypes.Pending),
+		oldTask:    oldTask,
```

Comment:
What is old an old task?

Reply:
old task struct. Im trying to limit changes in this PR to prevent having to change everything from top to bottom at once. Moving over to Ctrl requires you to use the new Task Struct, and not the one in internal/tasks.go. So this PR will result in both being used at once. Once its merged Ill throw up the next PR that will remove the oldTask and its code.